File: JasSwift/Sources/Interpreter/PointBuffers.swift
// Named point buffers — the Swift analogue of
// jas_dioxus/src/interpreter/point_buffers.rs.
//
// Used by tools that accumulate a sequence of (x, y) points during a
// drag and consume the whole list on mouseup (Lasso for polygonal
// selection, Pencil for freehand path fitting). The tool state scope
// (`$tool.<id>.*`) is scalar-oriented, so these buffers live outside
// the StateStore in module-local storage.
//
// Exposed to YAML via:
//   effect  buffer.push:  { buffer, x, y }
//   effect  buffer.clear: { buffer }
//   primitive buffer_length("<name>") -> number
//   effect  doc.select_polygon_from_buffer: { buffer, additive }
//   effect  doc.add_path_from_buffer: { buffer, fit_error? }
//   overlay render type `buffer_polygon` / `buffer_polyline`

import Foundation

// Thread-local storage — the Swift analogue of Rust's `thread_local!`
// in point_buffers.rs. A drag (press → move* → release) runs entirely
// on one thread (the main thread in the app), so the buffer persists
// across the gesture. A plain process-global `var` would instead be
// shared across threads, which corrupts the swift-testing concurrent
// thread pool: two tests using the same buffer name (e.g. "pencil")
// would clobber each other's points and produce empty/zeroed paths.
// Mirroring Rust's thread-local design isolates each thread's buffers.
private let _bufferKey = "jas.pointBuffers"

private func _bufferStore() -> NSMutableDictionary {
    let dict = Thread.current.threadDictionary
    if let existing = dict[_bufferKey] as? NSMutableDictionary {
        return existing
    }
    let fresh = NSMutableDictionary()
    dict[_bufferKey] = fresh
    return fresh
}

/// Append a point to the named buffer (creates it on first push).
func pointBuffersPush(_ name: String, _ x: Double, _ y: Double) {
    let store = _bufferStore()
    var buf = (store[name] as? [(Double, Double)]) ?? []
    buf.append((x, y))
    store[name] = buf
}

/// Empty the named buffer. No-op if it doesn't exist.
func pointBuffersClear(_ name: String) {
    _bufferStore().removeObject(forKey: name)
}

/// Number of points in the named buffer. 0 if missing.
func pointBuffersLength(_ name: String) -> Int {
    (_bufferStore()[name] as? [(Double, Double)])?.count ?? 0
}

/// Borrow the named buffer's points as an array. Empty if missing.
func pointBuffersPoints(_ name: String) -> [(Double, Double)] {
    (_bufferStore()[name] as? [(Double, Double)]) ?? []
}

File: jas_flask/static/js/engine/point_buffers.mjs
// Module-local point buffers for drag-accumulating tools.
//
// Tools like Pencil and Paintbrush push (x, y) coordinates here on
// each mousemove during a drag, then on mouseup the runtime reads
// the buffer (typically via doc.add_path_from_buffer) and clears it.
//
// JS port of workspace_interpreter/point_buffers.py. Single-threaded
// browser context; a module-local Map suffices.

const _buffers = new Map();

/** Reset the named buffer to empty. */
export function clear(name) {
  _buffers.set(name, []);
}

/** Append (x, y) to the named buffer. */
export function push(name, x, y) {
  let buf = _buffers.get(name);
  if (!buf) {
    buf = [];
    _buffers.set(name, buf);
  }
  buf.push([x, y]);
}

/** Number of points currently in the named buffer. */
export function length(name) {
  const buf = _buffers.get(name);
  return buf ? buf.length : 0;
}

/** Return a shallow copy of the buffer's points as [x, y] pairs. */
export function points(name) {
  const buf = _buffers.get(name);
  return buf ? buf.slice() : [];
}
